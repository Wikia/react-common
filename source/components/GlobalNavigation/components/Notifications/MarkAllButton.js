import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';

import NotificationsContext from '../../context/NotificationContext';

const MarkAllButton = ({ track, header }) => {
    const { markAllAsRead } = useContext(NotificationsContext);
    const onClick = useCallback(() => {
        track({
            action: 'click',
            category: 'on-site-notifications',
            label: 'mark-all-as-read',
        });

        markAllAsRead();
    }, [markAllAsRead, track]);

    return (
        <div className="wds-notifications__mark-all-as-read-button">
            {
                header && <p className="wds-notifications__header">{header}</p>
            }
            {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
            <a onClick={onClick} className="wds-notifications__mark-all-as-read">
                Mark all as read
            </a>
        </div>
    );
};

MarkAllButton.propTypes = {
    header: PropTypes.string.isRequired,
    track: PropTypes.func.isRequired,
};

export default MarkAllButton;
