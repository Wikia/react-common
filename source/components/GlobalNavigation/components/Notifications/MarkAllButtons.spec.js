import React, { useContext, useCallback } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme/build';
import merge from 'lodash/merge';

import MarkAllButton from './MarkAllButton';

const defaultProps = {
    header: '',
    track: () => null,
};

const mockDefaultContext = { markAllAsRead: () => null };

function renderComponent(props) {
    const renderer = new ShallowRenderer();
    const computedProps = merge({}, defaultProps, props);

    renderer.render(<MarkAllButton {...computedProps} />);

    return renderer.getRenderOutput();
}

// TODO: remove mocks when enzyme/react-test-renderer will fully support hooks
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(() => mockDefaultContext),
    useCallback: jest.fn(),
}));

/* eslint-disable-next-line jest/no-hooks */
beforeEach(() => {
    jest.resetModules();
});

test('MarkAllButton renders correctly with default props', () => {
    expect(renderComponent()).toMatchSnapshot();
});

test('MarkAllButton renders correctly with non-empty header', () => {
    expect(renderComponent({ header: 'blah' })).toMatchSnapshot();
});

test('MarkAllButton calls track and markAllAsReadMock onClick', () => {
    const markAllAsReadMock = jest.fn();
    const trackMock = jest.fn();
    useContext.mockImplementation(() => ({ ...mockDefaultContext, markAllAsRead: markAllAsReadMock }));
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({ track: trackMock });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('a').props().onClick();

    expect(markAllAsReadMock).toBeCalledWith();
    expect(trackMock).toBeCalledWith(expect.any(Object));
});
