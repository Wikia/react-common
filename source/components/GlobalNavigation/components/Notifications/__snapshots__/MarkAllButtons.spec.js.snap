// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkAllButton renders correctly with default props 1`] = `
<div
  className="wds-notifications__mark-all-as-read-button"
>
  
  <button
    className="wds-notifications__mark-all-as-read"
  >
    Mark all as read
  </button>
</div>
`;

exports[`MarkAllButton renders correctly with non-empty header 1`] = `
<div
  className="wds-notifications__mark-all-as-read-button"
>
  <p
    className="wds-notifications__header"
  >
    blah
  </p>
  <button
    className="wds-notifications__mark-all-as-read"
  >
    Mark all as read
  </button>
</div>
`;
